nmt.go

```diff
@@ -142,71 +149,119 @@ func (n *NamespacedMerkleTree) ProveRange(start, end int) (Proof, error) {
 func (n *NamespacedMerkleTree) ProveNamespace(nID namespace.ID) (Proof, error) {
 	isMaxNsIgnored := n.treeHasher.IsMaxNamespaceIDIgnored()
-	// In the cases (nID < minNID) or (maxNID < nID),
-	// return empty range and no proof:
-	if nID.Less(n.minNID) || n.maxNID.Less(nID) {
+	// case 1)
+	// In the cases (n.nID < minNID) or (n.maxNID < nID),
+	// return empty range and no proof
+	if nID.Less(n.minNID) || n.maxNID.Less(nID) { // TODO [Me] we could move this entire if block inside the `foundInRange` function
+		// TODO [Me] Shouldn't we instead return the first or the last node in the tree as the exclusion proof?
+		// TODO [Me] although I think this current logic is based on the premise that the root of the tree is trusted
 		return NewEmptyRangeProof(isMaxNsIgnored), nil
 	}
 
+	// find the range of indices of leaves with the given nID
 	found, proofStart, proofEnd := n.foundInRange(nID)
+
+	// case 2)
 	if !found {
 		// To generate a proof for an absence we calculate the
 		// position of the leaf that is in the place of where
 		// the namespace would be in:
-		proofStart = n.calculateAbsenceIndex(nID)
+		proofStart = n.calculateAbsenceIndex(nID) // TODO [Me] this could simply return a range, to avoid the line below
 		proofEnd = proofStart + 1
 	}
-	// At this point we either found the namespace in the tree or calculated
+
+	// case 3)
+	// At this point we either found leaves with the namespace nID in the tree or calculated
 	// the range it would be in (to generate a proof of absence and to return
 	// the corresponding leaf hashes).
-	n.computeLeafHashesIfNecessary()
+	n.computeLeafHashesIfNecessary()                 // TODO [Me] Why it is needed? cannot we make sure the leaves hashes are calculated as soon as a data is pushed to the tree?
 	proof := n.buildRangeProof(proofStart, proofEnd)
 
 	if found {
 		return NewInclusionProof(proofStart, proofEnd, proof, isMaxNsIgnored), nil
 	}
+	// TODO [Me] the underlying struct type for both inclusion and absence proofs are the same, i.e., `Proof`
+	// TODO [Me] the only distinction is in the presence/absence of the `leafHash`,
+	// TODO [Me] the interpretation of the `Proof` then may be a bit non-trivial
+	// TODO [Me] however, we may want to actually add an extra field to the `Proof`
+	// TODO [Me] that indicates whether the proof is for inclusion or for absence
+
 	return NewAbsenceProof(proofStart, proofEnd, proof, n.leafHashes[proofStart], isMaxNsIgnored), nil
 }
 
+// buildRangeProof returns the nodes (as byte slices) in the range proof of the supplied range i.e.,
+// [proofStart, proofEnd) where proofEnd is non-inclusive.
+// The nodes are ordered according to in order traversal of the namespaced tree. (TODO [Me] I need to double-check the order)
 func (n *NamespacedMerkleTree) buildRangeProof(proofStart, proofEnd int) [][]byte {
-	proof := [][]byte{}
+	// TODO a more secure way would be to make it slice of fixed size arrays i.e., the hash output size
```

Comment:
In that case, would we also return a slice of fixed size arrays? What if someone wants to use a different hash function (with a different output size)?

Reply:
If we expect varying namespace ID sizes, then nodes cannot be represented by a fixed size array.